### PR TITLE
Service metrics txn events txn metrics

### LINKED ISSUE #5243: service metrics: add numeric outcome
Add a numeric outcome to 
- transaction events (mapped as `byte`)
- transaction metrics (mapped as `byte` as it has an `event.outcome` dimension)
- service metrics (mapped as [`aggregate_metric_double`](https://www.elastic.co/guide/en/elasticsearch/reference/current/aggregate-metric-double.html), as it has no `event.outcome` dimension). 

All fields should be mapped as doc-value only fields.

The concrete field name needs to be defined. The working title is `event.outcome_numeric`.

This will help with reducing dimensions while keeping the structure of transaction metrics and service metrics similar. 

Mapping of `event.outcome` to `event.outcome_numeric`

| `event.outcome` | `event.outcome_numeric`|
|---|---|
|`unknown` | `null` |
| `failure` | 0 |
| `success` | 1 |

For the `aggregate_metric_double`, the following sub-fields should be populated:
- `sum`: the number of success events
- `value_count`: the number of success + error events

We can then simply do an average aggregation over the `event.outcome_numeric` to get the success rate. We need to validate that the avg aggregation works correctly on both the `byte` and `aggregate_metric_double` mapped fields.

---

Original request: 
Right now we use a terms aggregation or multiple filter aggregations on event.outcome to calculate transaction error rate. If we store failure as 0 and success in 1 in a single field (and not populate it for unknown), we can get the percentage by doing an avg aggregation on that single field. This would simplify our queries and the post processing that we are doing, will probably be a small performance improvement, and would make it easier for users to build custom dashboards.

### PR DESCRIPTION
**Pull request by carsonip**
## Motivation/summary

Add `event.success_count` to transaction events, transaction metrics and service metrics. It was previously available for service metrics at `transaction.success_count` by #9791 

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [x] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [x] Documentation has been updated

## How to test these changes

1. Check that `event.success_count` is correct for transaction event, transaction metrics and service metrics
2. Check that they work with avg aggregation

## Related issues

Closes #5243

### CODE REVIEW

**Comment by carsonip** [comment]
Keeping this in draft during review since it will have to go with https://github.com/elastic/apm-data/pull/2

File: apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml
Comment (axw) [inline apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml:19]: Seeing as this one only applies to trace events, should we define it inline here?

**Review by axw — COMMENTED** [review]
Looks great! Just one minor comment.

File: apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml
Comment (carsonip) [reply apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml:19]: it was put into a named pipeline because I didn't know whether this should be used in apm_data_stream_migration as well. If it only applies to traces default, then yes it is better to inline it

Comment (carsonip) [reply apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml]: done

**Review by axw — APPROVED** [review]

File: apmpackage/apm/data_stream/internal_metrics/manifest.yml
Comment (simitt) [inline apmpackage/apm/data_stream/internal_metrics/manifest.yml:29]: It should be mapped as `doc-value-only` (setting `index: false`).

File: x-pack/apm-server/aggregation/txmetrics/aggregator.go
Comment (simitt) [inline x-pack/apm-server/aggregation/txmetrics/aggregator.go:400]: Is it necessary to use `float64()` here?

File: changelogs/head.asciidoc
Comment (simitt) [inline changelogs/head.asciidoc:14]: The changelog is missing that `event.success_count` has been introduced also for transaction events and transaction metrics. This changelog should contain all changes made to APM Server.

File: apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml
Comment (simitt) [inline apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml:26]: I'd also add `"ignore_failure": true`.

**Review by simitt — COMMENTED** [review]
Code mostly looks good, left a couple minor comments.

File: apmpackage/apm/data_stream/internal_metrics/manifest.yml
Comment (carsonip) [reply apmpackage/apm/data_stream/internal_metrics/manifest.yml:29]: `aggregate_metric_double` does not support `index: false` as mentioned in https://github.com/elastic/apm-server/issues/5243#issuecomment-1351349456

Comment (simitt) [reply apmpackage/apm/data_stream/internal_metrics/manifest.yml:29]: ah I missed that, thanks for pointing out.

File: x-pack/apm-server/aggregation/txmetrics/aggregator.go
Comment (carsonip) [reply x-pack/apm-server/aggregation/txmetrics/aggregator.go:400]: Unfortunately since we are reusing SummaryMetric here, the Sum is a float64 and it will not compile when it is assigned an int64
```
# github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics
x-pack/apm-server/aggregation/txmetrics/aggregator.go:400:27: cannot use totalCount (variable of type int64) as type float64 in assignment
make: *** [Makefile:61: build/apm-server-linux-amd64] Error 2
```

File: changelogs/head.asciidoc
Comment (carsonip) [reply changelogs/head.asciidoc:9]: Updated changelog, but split the part about transaction events and transaction metrics outside "breaking changes".

File: apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml
Comment (carsonip) [reply apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml:27]: Added. But it isn't obvious under what circumstances this will fail, given the null event check. That is why I didn't add `ignore_failure` in the first place. Do you have any hints?

Comment (simitt) [reply apmpackage/apm/data_stream/traces/elasticsearch/ingest_pipeline/default.yml:27]: agreed, no obvious circumstance, but not having it did bite us in the past, so we basically have it on all processors.

**Comment by simitt** [comment]
@carsonip the PR looks pretty solid - are you planning on marking it ready for review soon?

**Comment by carsonip** [comment]
> @carsonip the PR looks pretty solid - are you planning on marking it ready for review soon?

Was trying to squash locally before making it ready for review but got into some local permission errors which I thought we fixed yesterday. Anyway, can do a squash merge after approval.

**Review by marclop — APPROVED** [review]

**Review by simitt — APPROVED** [review]

**Comment by simitt** [comment]
Merged, to unblock @marclop who is working on a PR touching the same code base.

**Comment by carsonip** [comment]
Thanks @simitt

**Comment by lahsivjar** [comment]
Tested with `8.7.0-BC4`:

<details>
<summary>Check that event.success_count is correct for transaction event, transaction metrics and service metrics</summary>

### Transaction Events

Asserted that the metric is recorded as `byte`.
<details>
<summary>Success</summary>

```json
GET /traces-apm-default*/_search?filter_path=hits.hits
{
  "size": 1,
  "query": {
    "bool": {
      "should": [
        {"match": {"event.outcome": "success"}},
        {"match": {"service.name": "apm2"}}
      ]
    }
  }, 
  "_source": [false],
  "fields": [
    "event.success_count"
  ]
}
```
```json
{
  "hits": {
    "hits": [
      {
        "_index": ".ds-traces-apm-default-2023.03.06-000001",
        "_id": "flwTtYYBNtUXE48ncG4T",
        "_score": 4.617365,
        "_source": {},
        "fields": {
          "event.success_count": [
            1
          ]
        }
      }
    ]
  }
}
```
</details>

<details>
<summary>Failure</summary>

```json
GET /traces-apm-default*/_search?filter_path=hits.hits
{
  "size": 1,
  "query": {
    "bool": {
      "should": [
        {"match": {"event.outcome": "failure"}},
        {"match": {"service.name": "apm2"}}
      ]
    }
  }, 
  "_source": [false],
  "fields": [
    "event.success_count"
  ]
}
```
```json
{
  "hits": {
    "hits": [
      {
        "_index": ".ds-traces-apm-default-2023.03.06-000001",
        "_id": "gFwTtYYBNtUXE48ncG4T",
        "_score": 9.915409,
        "_source": {},
        "fields": {
          "event.success_count": [
            0
          ]
        }
      }
    ]
  }
}
```
</details>

### Transaction Metrics

Asserted that the metric is recorded as `aggregate_metric_double` with `sum` and `value_count` metrics.
<details>
<summary>Success</summary>

```json
GET /metrics-apm.transaction*/_search?filter_path=hits.hits
{
  "size": 1,
  "query": {
    "bool": {
      "should": [
        {"match": {"event.outcome": "success"}},
        {"match": {"service.name": "apm2"}}
      ]
    }
  }, 
  "_source": [false],
  "fields": [
    "event.success_count"
  ]
}
```
```json
{
  "hits": {
    "hits": [
      {
        "_index": ".ds-metrics-apm.transaction.1m-default-2023.03.06-000001",
        "_id": "aFwUtYYBNtUXE48nDm9C",
        "_score": 0.69894236,
        "_source": {},
        "fields": {
          "event.success_count": [
            {
              "sum": 2,
              "value_count": 2
            }
          ]
        }
      }
    ]
  }
}
```
</details>
<details>
<summary>Failure</summary>

```json
GET /metrics-apm.transaction*/_search?filter_path=hits.hits
{
  "size": 1,
  "query": {
    "bool": {
      "should": [
        {"match": {"event.outcome": "failure"}},
        {"match": {"service.name": "apm2"}}
      ]
    }
  }, 
  "_source": [false],
  "fields": [
    "event.success_count"
  ]
}
```
```json
{
  "hits": {
    "hits": [
      {
        "_index": ".ds-metrics-apm.transaction.1m-default-2023.03.06-000001",
        "_id": "alwUtYYBNtUXE48nDm9C",
        "_score": 0.7566507,
        "_source": {},
        "fields": {
          "event.success_count": [
            {
              "sum": 0,
              "value_count": 2
            }
          ]
        }
      }
    ]
  }
}
```
</details>

### Service Transaction Metrics

Asserted that the metric is recorded as `aggregate_metric_double` with `sum` and `value_count` metrics.
<details>
<summary>Success</summary>

```json
GET /metrics-apm.service_transaction*/_search?filter_path=hits.hits
{
  "size": 1,
  "query": {
    "bool": {
      "should": [
        {"match": {"event.outcome": "success"}},
        {"match": {"service.name": "apm2"}}
      ]
    }
  }, 
  "_source": [false],
  "fields": [
    "event.success_count"
  ]
}
```
```json
```
</details>
<details>
<summary>Failure</summary>

```json
GET /metrics-apm.service_transaction*/_search?filter_path=hits.hits
{
  "size": 1,
  "query": {
    "bool": {
      "should": [
        {"match": {"event.outcome": "failure"}},
        {"match": {"service.name": "apm2"}}
      ]
    }
  }, 
  "_source": [false],
  "fields": [
    "event.success_count"
  ]
}
```
```json
{
  "hits": {
    "hits": [
      {
        "_index": ".ds-metrics-apm.service_transaction.1m-default-2023.03.06-000001",
        "_id": "fFwUtYYBNtUXE48nDm9C",
        "_score": 0.03423317,
        "_source": {},
        "fields": {
          "event.success_count": [
            {
              "sum": 0,
              "value_count": 2
            }
          ]
        }
      }
    ]
  }
}
</details>

<details>
<summary> Check that they work with avg aggregation </summary>

### Transaction Events
```json
GET /traces-apm-default*/_search?filter_path=aggregations
{
  "size": 1,
  "aggs": {
    "err_rate": {
      "filter": {
        "term": {
          "service.name": "apm2"
        }
      },
      "aggs": {
        "err_rate": {"avg": { "field": "event.success_count" }}
      }
    }
  }
}
```
```json
{
  "aggregations": {
    "err_rate": {
      "doc_count": 200,
      "err_rate": {
        "value": 0.5
      }
    }
  }
}
```

### Transaction Metrics
```json
GET /metrics-apm.transaction*/_search?filter_path=aggregations
{
  "size": 1,
  "aggs": {
    "err_rate": {
      "filter": {
        "term": {
          "service.name": "apm2"
        }
      },
      "aggs": {
        "err_rate": {"avg": { "field": "event.success_count" }}
      }
    }
  }
}
```
```json
{
  "aggregations": {
    "err_rate": {
      "doc_count": 200,
      "err_rate": {
        "value": 0.5
      }
    }
  }
}
```

### Service Transaction Metrics
```json
GET /metrics-apm.service_transaction*/_search?filter_path=aggregations
{
  "size": 1,
  "aggs": {
    "err_rate": {
      "filter": {
        "term": {
          "service.name": "apm2"
        }
      },
      "aggs": {
        "err_rate": {"avg": { "field": "event.success_count" }}
      }
    }
  }
}
```
```json
{
  "aggregations": {
    "err_rate": {
      "doc_count": 200,
      "err_rate": {
        "value": 0.5
      }
    }
  }
}
```
</details>

<details>

<summary> Ensure event.success_count.sum is rounded under sampling setup</summary>


</details>

**Pull request merged**
